gzochi-server/doc/gzochi-metad.8
.TH GZOCHI-METAD 8 "April 1, 2016"
.SH NAME
.Nm gzochi-metad - the gzochi meta server
.SH SYNOPSIS
\fBgzochi-metad [\fR\fB\fIoptions\fR\fB]\fR
.SH DESCRIPTION
\fBgzochi-metad\fR, the meta server for the gzochi massively multiplayer online
game development framework, is responsible for supporting gzochi game
application servers in a high-availability game server cluster with data and
coordination services. As its name suggests, it's a server for servers.
.P
When gzochi-metad starts, it begins listening for connections from gzochid
processes, and optionally starts a web server that reports information about the
state of the game server cluster.
.P
The \fBgzochi-metad\fR program has many options that can be specified at
startup. For a complete list of options, run this command:
.IP
shell> \fBgzochi-metad \-\-help\fR
.SH ENVIRONMENT
.TP
.B GZOCHI_METAD_CONF_LOCATION
If \fB$GZOCHI_METAD_CONF_LOCATION\fR is set before \fBgzochi-metad\fR is
started, its value is used as the full path to the meta server configuration
file. This variable will be overriden by a command-line argument giving the
location of the file, if present.
.SH SEE ALSO
.P
The full documentation for \fBgzochi-metad\fR is maintained as a Texinfo
manual. If the \fBinfo\fR and \fBgzochi-metad\fR programs are properly 
installed at your site, the command
.IP
.B info gzochi-metad
.P
should give you access to the complete manual.
.SH COPYRIGHT
Copyright \(co 2016 Julian Graham. License GPLv3+: GNU GPL version 3
or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH AUTHOR
Julian Graham (joolean@undecidable.net)


NEWS
gzochi NEWS -- history of user-visible changes
Copyright (C) 2015 Julian Graham
See the end of copying conditions.

Send comments, questions, and bug reports to gzochi-devel@nongnu.org

Changes in 0.8 (since 0.7)

* Changes to the distribution

** Removed deprecated synchronous client (libgzochi)

libgzochi, the synchronous reference client implementation included with earlier
versions of the gzochi framework, has been removed from the distribution. The
client was deprecated in an earlier release, since the concurrency strategy
suggested by its API was undesirable. Application developers are encouraged to
use libgzochi-glib instead.

** Removed client support static library (libgzochi-client-common)

libgzochi-client-common, the static library that facilitated sharing of code
between libgzochi and libgzochi-glib, has been merged with libgzochi-glib and no
longer exists as an independent build target within the gzochi framework source
distribution.

** New example: External transactions

Client and server code for a new example application has been added to the 
source distribution under the "examples/" tree. This example demonstrates 
techniques for synchronizing a transaction executed within the gzochid container
with an external transaction-aware process, such as a relational database.


Changes in 0.7 (since 0.6)

* Changes to the distribution

** New reference client library implementation: (gzochi client)

Game developers who want to keep their entire application stack in Guile Scheme
(C server container notwithstanding) are in luck! The new (gzochi client)
library is a Scheme reference client implementation with an execution model that
can be integrated into a main loop or used stand-alone. The source distribution
for (gzochi client) is under the "gzochi-guile-client/" tree.


Changes in 0.6 (since 0.5)

* Compatibility

** Support for additional platforms

gzochi now builds on Mac OS X and (likely) other BSD variants.


Changes in 0.3 (since 0.2)

* Changes to the distribution

** New client support static library: libgzochi-client-common

Some of the trickier aspects of client communication with a gzochi server
instance, such as maintaining a buffer for partial message delivery, have been
factored out of libgzochi and into a static library, with a source distribution
under the "gzochi-client-common/" tree, that can be built into other client 
implementations. 

** New reference client library implementation: libgzochi-glib

libgzochi-client-common already has a consumer in the distribution:
libgzochi-glib, a new reference client implementation that integrates easily 
with the "main loop" functionality in the popular GNU GLib library to support
the development of single-threaded client applications. The source distribution
for libgzochi-glib is under the "gzochi-glib-client/" tree.

** Removed deprecated debugging client

mictlan, the primitive "debugging client" included with earlier versions of the
gzochi framework, has been removed from the distribution. Its functionality has
been superceded by the debugging capabilities of gzochid itself, which can be 
accessed using a telnet client.


Changes in 0.2 (since 0.1)

* Changes to the distribution

** New example: abermud

Client and server code for a new example game -- based on Alan Cox's 
AberMUD -- has been added to the source distribution under the "examples/"
tree. This example demonstrates techniques for maintaining a persistent world
in which state is shared between multiple players.


Copyright information:

Copyright (C) 2015 Julian Graham

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


gzochi-server/src/gzochi-metad.c
/* gzochi-metad.c: Main server bootstrapping routines for gzochi-metad
 * Copyright (C) 2016 Julian Graham
 *
 * gzochi is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <getopt.h>
#include <glib.h>
#include <glib-object.h>
#include <libintl.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "config.h"
#include "dataserver.h"
#include "httpd.h"
#include "httpd-meta.h"
#include "log.h"
#include "resolver.h"
#include "socket.h"

#define _(String) gettext (String)

#define Q(x) #x
#define QUOTE(x) Q(x)

#ifndef GZOCHI_METAD_CONF_LOCATION
#define GZOCHI_METAD_CONF_LOCATION "/etc/gzochi-metad.conf"
#endif /* GZOCHI_METAD_CONF_LOCATION */

#define GZOCHI_METAD_TYPE_ROOT_CONTEXT gzochi_metad_root_context_get_type ()

/* Boilerplate setup for the gzochi-metad root context. */

/* The following boilerplate can be consolidated once GLib 2.44 makes it into
   Debian stable and `G_DECLARE_FINAL_TYPE' can be used. */

GType gzochi_metad_root_context_get_type (void);

struct _GzochiMetadRootContextClass
{
  GObjectClass parent_class;
};

typedef struct _GzochiMetadRootContextClass GzochiMetadRootContextClass;

/* The root context object. */

struct _GzochiMetadRootContext
{
  GObject parent_intance;

  GzochidHttpServer *http_server; /* The admin web console. */

  /* The meta server global socket server. */

  GzochidSocketServer *socket_server;
  
  GzochiMetadDataServer *data_server; /* The meta server data server. */
};

typedef struct _GzochiMetadRootContext GzochiMetadRootContext;

static inline GzochiMetadRootContext *
GZOCHI_METAD_ROOT_CONTEXT (gconstpointer ptr) {
  return G_TYPE_CHECK_INSTANCE_CAST
    (ptr, gzochid_configuration_get_type (), GzochiMetadRootContext);
}

G_DEFINE_TYPE (GzochiMetadRootContext, gzochi_metad_root_context,
	       G_TYPE_OBJECT);

enum gzochi_metad_root_context_properties
  {
    PROP_HTTP_SERVER = 1,
    PROP_SOCKET_SERVER,
    PROP_DATA_SERVER,
    N_PROPERTIES
  };

static GParamSpec *obj_properties[N_PROPERTIES] = { NULL };

static void
root_context_set_property (GObject *object, guint property_id,
			   const GValue *value, GParamSpec *pspec)
{
  GzochiMetadRootContext *self = GZOCHI_METAD_ROOT_CONTEXT (object);

  switch (property_id)
    {
    case PROP_HTTP_SERVER:
      self->http_server = g_object_ref (g_value_get_object (value));
      break;

    case PROP_SOCKET_SERVER:
      self->socket_server = g_object_ref (g_value_get_object (value));
      break;

    case PROP_DATA_SERVER:
      self->data_server = g_object_ref (g_value_get_object (value));
      break;
      
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
      break;
    }
}

static void
gzochi_metad_root_context_class_init (GzochiMetadRootContextClass *klass)
{
  GObjectClass *object_class = G_OBJECT_CLASS (klass);

  object_class->set_property = root_context_set_property;
  obj_properties[PROP_HTTP_SERVER] = g_param_spec_object
    ("http-server", "Admin HTTP server", "Set the admin HTTP server",
     GZOCHID_TYPE_HTTP_SERVER,
     G_PARAM_WRITABLE | G_PARAM_CONSTRUCT | G_PARAM_PRIVATE);

  obj_properties[PROP_SOCKET_SERVER] = g_param_spec_object
    ("socket-server", "Socket server", "Set the socket server",
     GZOCHID_TYPE_SOCKET_SERVER,
     G_PARAM_WRITABLE | G_PARAM_CONSTRUCT | G_PARAM_PRIVATE);

  obj_properties[PROP_DATA_SERVER] = g_param_spec_object
    ("data-server", "Data server", "Set the data server",
     GZOCHI_METAD_TYPE_DATA_SERVER,
     G_PARAM_WRITABLE | G_PARAM_CONSTRUCT | G_PARAM_PRIVATE);

  g_object_class_install_properties
    (object_class, N_PROPERTIES, obj_properties);
}

static void
gzochi_metad_root_context_init (GzochiMetadRootContext *self)
{
}

static const struct option longopts[] =
  {
    { "config", required_argument, NULL, 'c' },
    { "help", no_argument, NULL, 'h' },
    { "version", no_argument, NULL, 'v' },
    { NULL, 0, NULL, 0 }
  };

/* Prints the version and copyright statement to standard output. */

static void
print_version (void)
{
  printf ("gzochi-metad (gzochi) %s\n", VERSION);

  puts ("");
  printf (_("\
Copyright (C) %s Julian Graham\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n"),
	  "2016");
}

/* Prints help and configuration options to standard output. */

static void
print_help (const char *program_name)
{
  printf (_("\
Usage: %s [OPTION]...\n"), program_name);
  
  puts ("");
  fputs (_("\
  -c, --config        full path to gzochi-metad.conf\n\
  -h, --help          display this help and exit\n\
  -v, --version       display version information and exit\n"), stdout);

  puts ("");
  printf (_("\
Report bugs to: %s\n"), PACKAGE_BUGREPORT);
#ifdef PACKAGE_PACKAGER_BUG_REPORTS
  printf (_("Report %s bugs to: %s\n"), PACKAGE_PACKAGER,
          PACKAGE_PACKAGER_BUG_REPORTS);
#endif /* PACKAGE_PACKAGER_BUG_REPORTS */

#ifdef PACKAGE_URL
  printf (_("%s home page: <%s>\n"), PACKAGE_NAME, PACKAGE_URL);
#else
  printf (_("%s home page: <http://www.nongnu.org/%s/>\n"),
          PACKAGE_NAME, PACKAGE);
#endif /* PACKAGE_URL */
}

/* Set the logging threshold. */

static void
initialize_logging (GKeyFile *key_file)
{
  GHashTable *log_config =
    gzochid_config_keyfile_extract_config (key_file, "log");

  if (g_hash_table_contains (log_config, "priority.threshold")) 
    {
      char *threshold = g_hash_table_lookup (log_config, "priority.threshold");

      if (g_ascii_strncasecmp (threshold, "EMERG", 5) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_ERROR);
      else if (g_ascii_strncasecmp (threshold, "ALERT", 5) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_ERROR);
      else if (g_ascii_strncasecmp (threshold, "CRIT", 4) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_ERROR);
      else if (g_ascii_strncasecmp (threshold, "ERR", 3) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_CRITICAL);
      else if (g_ascii_strncasecmp (threshold, "WARNING", 7) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_WARNING);
      else if (g_ascii_strncasecmp (threshold, "NOTICE", 6) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_MESSAGE);
      else if (g_ascii_strncasecmp (threshold, "INFO", 4) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_INFO);
      else if (g_ascii_strncasecmp (threshold, "DEBUG", 5) == 0)
        gzochid_install_log_handler (G_LOG_LEVEL_DEBUG);
    }
  else gzochid_install_log_handler (G_LOG_LEVEL_INFO);

  g_hash_table_destroy (log_config);
}

/* Configure and start the admin web console if it's enabled. */

static void
initialize_httpd (GzochidHttpServer *http_server, GKeyFile *key_file)
{
  GHashTable *admin_config =
    gzochid_config_keyfile_extract_config (key_file, "admin");

  if (gzochid_config_to_boolean
      (g_hash_table_lookup (admin_config, "module.httpd.enabled"), FALSE))
    {
      int port = gzochid_config_to_int
	(g_hash_table_lookup (admin_config, "module.httpd.port"), 8800);

      gzochid_httpd_meta_register_handlers (http_server);
      gzochid_http_server_start (http_server, port, NULL);
    }
  
  g_hash_table_destroy (admin_config);
}

int 
main (int argc, char *argv[])
{
  GError *err = NULL;
  
  const char *program_name = argv[0];
  const char *conf_path = NULL;
  int optc = 0;

  GKeyFile *key_file = g_key_file_new ();
  GzochidConfiguration *configuration = NULL;
  GzochidResolutionContext *resolution_context = g_object_new
    (GZOCHID_TYPE_RESOLUTION_CONTEXT, NULL);
  GzochiMetadRootContext *root_context = NULL;
  
  setlocale (LC_ALL, "");

  while ((optc = getopt_long (argc, argv, "c:hv", longopts, NULL)) != -1)
    switch (optc)
      {
      case 'c':
	conf_path = strdup (optarg);
	break;
      case 'v':
	print_version ();
	exit (EXIT_SUCCESS);
	break;
      case 'h':
	print_help (program_name);
	exit (EXIT_SUCCESS);
	break;
      case '?': exit (EXIT_FAILURE);
      }

  if (conf_path == NULL)
    {
      const char *env = getenv ("GZOCHI_METAD_CONF_LOCATION");
      conf_path = env ? env : QUOTE(GZOCHI_METAD_CONF_LOCATION);
    }

  g_info ("Reading configuration from %s", conf_path);
  g_key_file_load_from_file (key_file, conf_path, G_KEY_FILE_NONE, &err);
  
  if (err != NULL)
    {
      /*
	The server has reasonable defaults for everything in gzochi-metad.conf,
	so this condition is not fatal.
      */

      g_warning 
        ("Failed to load server configuration file %s: %s", conf_path,
	 err->message);
      g_clear_error (&err);
    }

  initialize_logging (key_file);
  
  configuration = g_object_new
    (GZOCHID_TYPE_CONFIGURATION, "key_file", key_file, NULL);
  gzochid_resolver_provide (resolution_context, G_OBJECT (configuration), NULL);
  
  root_context = gzochid_resolver_require_full
    (resolution_context, GZOCHI_METAD_TYPE_ROOT_CONTEXT, &err);

  if (err != NULL)
    {
      g_error ("Failed to bootstrap gzochi-metad: %s", err->message);
      exit (1);
    }

  initialize_httpd (root_context->http_server, key_file);
  gzochi_metad_dataserver_start (root_context->data_server);
  g_main_loop_run (root_context->socket_server->main_loop);
  
  return 0;
}
